lib/actions/ssh-direct.ts

```diff
@@ -0,0 +1,245 @@
+			resolve(socket);
+		};
+
+		socket
+			.on('end', earlyEnd)
+			.on('error', earlyError)
+			.on('data', proxyData);
+	});
+
+Bluebird.promisifyAll(ssh2.Client);
+
+interface Ssh2ClientAsync extends ssh2.Client {
+	execAsync(
+		cmd: string,
+		options?: ssh2.ExecOptions,
+	): Bluebird<ssh2.ClientChannel>;
+}
+
+const createSSHClient = (socket: stream.Duplex): Bluebird<Ssh2ClientAsync> =>
+	new Bluebird((resolve, reject) => {
+		const client = new ssh2.Client() as Ssh2ClientAsync;
+		return client
+			.on('ready', () => {
+				resolve(client);
+			})
+			.on('error', err => {
+				const errSource = err && err.level ? 'ssh client socket' : 'ssh client';
+				let errMsg = `${errSource} error while initiating SSH connection`;
+				if (err && err.description) {
+					errMsg += `: ${err.description}`;
+				}
+				console.error(errMsg);
+				reject(new Error(errMsg));
+			})
+			.connect({
+				sock: socket,
+				username: 'root',
+				port: 22222,
+				agent: process.env.SSH_AUTH_SOCK,
+			});
+	});
+
+const createSSHClientDisposer = (
+	socket: stream.Duplex,
+): Bluebird.Disposer<Ssh2ClientAsync> =>
+	createSSHClient(socket).disposer((client: Ssh2ClientAsync) => client.end());
+
+interface ExecResponse {
+	code: number;
+	data?: {
+		stdout: string;
+		stderr: string;
+	};
+	signal?: string;
+}
+
+const pipeDeviceToStdout = (
+	deviceStream: ssh2.ClientChannel,
+): Bluebird<ExecResponse> => {
+	const dsResize = () =>
+		deviceStream.setWindow(process.stdout.rows!, process.stdout.columns!, 0, 0);
+	const dsEnd = () => deviceStream.end();
+	const dsWrite = (data: Buffer) => deviceStream.write(data);
+
+	if (process.stdin != null && process.stdout.isTTY) {
+		if (typeof process.stdin.setRawMode === 'function') {
+			process.stdin.setRawMode(true);
+		}
+
+		// Set initial window size and handle resize events
+		dsResize();
+		process.stdout.on('resize', dsResize);
+	}
+
+	return new Bluebird((resolve, reject) => {
+		const errorHandler = (err: Error) => {
+			console.error(err);
+			reject(err);
+		};
+
+		deviceStream
+			.on('data', (data: string) => process.stdout.write(data))
+			.on('end', () => {
+				process.stdin.removeListener('data', dsWrite);
+				process.stdin.removeListener('end', dsEnd);
+				process.stdin.removeListener('close', dsEnd);
+			})
+			.on('close', (code: number, signal: string) => resolve({ code, signal }))
+			.stderr.on('data', (data: string) => process.stderr.write(data))
+			.on('error', errorHandler);
+
+		process.stdin
+			.on('data', dsWrite)
+			// make sure that the device-side stream is properly closed if userStream ends first
+			.on('end', dsEnd)
+			.on('close', dsEnd)
+			.on('error', errorHandler);
+	});
+};
+
+const createCommandPipe = (param: {
+	cmd: string;
+	socket: net.Socket;
+}): Bluebird<ExecResponse> =>
+	Bluebird.using(createSSHClientDisposer(param.socket), client =>
+		client
+			.execAsync(param.cmd, { pty: process.stdout.isTTY })
+			.then(pipeDeviceToStdout),
```

Comment:
Not sure where exactly where the issue is, but testing the recent Charlie's Tips 'n' Tricks from [balena's newsletter](https://www.balena.io/blog/balena-monthly-roundup-january-2020/):

![Charlie's Tips 'n' Tricks](https://assets.balena.io/newsletter/2020-01/pipe.png)

I have found an incompatibility with the `-x` alternative. Note how, without `-x`, the commands and the shell prompt get printed to the console, whereas with `-x` they don't:

```
$ echo "uptime; exit;" | ./bin/balena-dev ssh da4da236b4f76820d69941e315fd215a
root@da4da23:~# uptime; exit;
 02:52:56  up  12:05,  1 user,  load average: 0.27, 0.24, 0.18
logout
```
```
$ echo "uptime; exit;" | ./bin/balena-dev ssh da4da236b4f76820d69941e315fd215a -x
=============================================================
    Welcome to balenaOS
=============================================================
 02:45:36  up  11:57,  0 users,  load average: 0.12, 0.15, 0.14
```

Reply:
Good point, I will add this in. Should I add a `-s`, `-q`, or similar option to suppress the header also?